man/genl-ctrl-list.8
.TH genl\-ctrl-list 8 "20 April 2012" "libnl"
.SH NAME
genl\-ctrl\-list \- List available kernel-side generic netlink families
.SH SYNOPSIS
.B genl\-ctrl\-list [-d]

.SH DESCRIPTION
.PP
Queries the generic netlink controller in kernel and prints a list of all
registered generic netlink families including the version of the interface
that has been registered.

.SH OPTIONS
.TP
.BR \-\^h " or " \-\-help
Print help text to console and exit.
.TP
.BR \-\^v " or " \-\-version
Print versioning information to console and exit.
.TP
.BR \-\^d " or " \-\-details
Include additional detailed information for each generic netlink
family that is printed.

The information includes:
.RS
.IP \(bu
The size of the family specific netlink message header (hdrsize).
.IP \(bu
The maximum netlink attribute identifier allowed by the interface
(maxattr).
.IP \(bu
A list of registered generic netlink operations prefixed with op
including their name (if available), identifier and the flags
indicating the availability of a doit or dump function.
.IP \(bu
A list of registered multicast groups prefixed with grp including
their name (if available) and identifier.
.RE

.RS
Example:

0x0010 nlctrl version 2
.RS 0
    hdrsize 0 maxattr 7
.RS 0
      op GETFAMILY (0x03) <has-doit,has-dump,has-policy>
.RS 0
      grp notify (0x10)
.RE


.SH EXAMPLE
.RS 0
$ genl-ctrl-list 
.RS 0
0x0010 nlctrl version 2
.RS 0
0x0011 NLBL_MGMT version 3
.RS 0
0x0012 NLBL_CIPSOv4 version 3
.RS 0
0x0013 NLBL_UNLBL version 3
.RS 0
0x0014 acpi_event version 1
.RS 0
0x0015 thermal_event version 1
.RS 0
0x0016 VFS_DQUOT version 1
.RS 0
0x0017 TASKSTATS version 1
.RS 0
0x0018 NET_DM version 2

.SH AUTHOR
.PP
Thomas Graf is the original author and current maintainer of libnl and
libnl tools. Many people have contributed to it since.


ChangeLog
Summary of Changes from 1.0-pre6 to HEAD
================================================
	Thomas Graf <tgraf@suug.ch>
	  o Generic netlink support
	  o Route Addition/Deletion
	  o Added nl_cache_subset()
	  o Have nl_object_clone() make real clones without
	    sharing any data.
	  o Remove old nl_object_alloc() not based on a object
	    operations to avoid bugs due to missing init.
	  o Added nl-list-caches utility
	  o Removed nlmsg_build_no_hdr(), duplicate
	  o Reworked message interface
	  o Fixed nlmsg_put() and genlmsg_put() to correctly reserve
	    tail room for user specific headers.
	  o Added nl_cache_move()
	  o Renamed nl_cache_delete() to nl_cache_remove() (no API break)
	  o Fixed reference counting while objects stay in caches.
	  o Object marking
	  o Moved attribute mask for objects into generic structure
	  o nl-list-caches: List available dump procedures
	  o Use PAGE_SIZE as initial buffer size when reading from
	    netlink socket
	  o Double buffer size when recv() returns MSG_TRUNC
	  o Replaced filter object operation with new compare operation
	    capable of listing differences between two objects
	  o Added nl_object_identical() to check if two objects are
	    identical from a uniqueness point of view
	  o Added nl_object_diff() returning a bitmask of differences in
	    attributes
	  o Added nl_object_attr_list() generating a list of attribute
	    name the object has assigned 
	  o Cache updates based on event notifications, code based on
	    Patrick McHardy's patches
	  o Cache Manager
	  o Added NL_AUTO_PID, NL_AUTO_SEQ for convenience
	  o Disable MSG_PEEK by default and provide nl_socket_enable_msg_peek()
	  o Fixed nl_recvmsgs() to return 0 when interrupted via NL_STOP or
	    NL_SKIP
	  o Fixed nl_recvmsgs() to stop reading after parsing if not in the
	    middle of a multipart message.
	  o Fixed nl_recvmsgs() to not stop after receving an ACK
	  o Fixed nl_recvmsgs() to not blindly discard remaining messages
	    if a NLMSG_DONE message is received.
	
	Petr Gotthard <petr.gotthard@siemens.com>
	Siemens AG Oesterreich
	  o Fix u32 to properly handle multiple keys
	  o rtnl_htb_set_(r|c)buffer()
	  o Fixed MTU handling in HTB class, problem pointed out
	    by Edouard Thuleau
	
	Zilvinas Valinskas <zilvinas@wilibox.com>
	  o Fix wrong msg_namelen in nl_recv()
	  o Fix memory leak in flnl_request_put()

	Helmut Schaa <hschaa@suse.de>
	  o Fix for using libnl from within C++

	Patrick McHardy <kaber@trash.net>
	  o *_alloc_cache(): Only refill cache if handle is provided
	
	James Oakley <jfunk@funktronics.ca>
	  o Fix rtnl_link_set_arptype() typo

	Philip Craig <philipc@snapgear.com>
	  o Change address family type from char to int
	  o Fix the error handling when the build fails.
	  o add nl_cache_mngr_get_fd()
	  o add netfilter support
	  o add netfilter conntrack support
	  o add netfilter log support

Summary of Changes from 1.0-pre5 to 1.0-pre6
================================================
	Christopher Aillon <caillon@redhat.com>
	  o Use $(libdir) instead of $(prefix)/lib for 64bit awesomeness.

	Thomas Graf <tgraf@suug.ch>
	  o Extend nl_msg to include source address, destination address
	    and the protocol being used.
	  o Make nl_send*() take a nl_msg instead of a nlmsghdr (API BREAK)
	  o Change callbacks to take a nl_msg instead of source address
	    and nlmsghdr (API BREAK)
	  o caches must specify the protocol they're hooked up from now on
	    if they intend to be associated with message types.
	  o cache_mngt_associate now takes the protocol besides the message
	    type to allow for multiple protocols to be supported (API BREAK)
	  o overwrite destination address in nl_send() when specified in the
	    message itself, allows for unbound addressing.
	  o Support for netlink based fib_lookup()
	  o Documentation fixes
	  o Fix double nlmsg_free() in nl_recvmsgs() while receiving
	    a multipart message and the read was interrupted.
	  o Change cache operations to store names for message types.
	  o Provide interface to convert message type to character string.
	  o Add dp_dump_msgtype to prefix each dumped element with a
	    pretty printed message type.
	  o netlink fib lookup support
	  o nl_recvmsgs() debugging
	  o use nl_cachemngt_type2name() when pretty printing netlink header
	  o Routing protocol translations.
	  o Routing metric translations.
	  o Revised route dumping
	  o Nexthop flag translations.
	  o Add support for IFF_DORMANT

	Petr Gotthard <petr.gotthard@siemens.com>
	Siemens AG Oesterreich
	  o Fix access to obj after freeing it
	  o Fix u32 selector access after realloc() 
	  o Fix missing out-of-memory error handling in various places
	  o Enhance nl-monitor to have group selection selectable and
	    demonstrate usage of select()
	  o Don't ignore IFF_RUNNING any longer
	  o fw classifier support
	
	Patrick McHardy <kaber@trash.net>
	  o Fix conflicting types for __u64
	  o Fix printf format string warnings
	  o Fix object cloning
	  o Deal with structure padding in nl_object_clone
	  o Fix nl_addr leak
	  o Set ce_msgtype in all parsed objects
	  o Fix addr flag filter
	  o Fix RTNLGRP definitions (was based on broken kernel version)
	  o Export nl_get_errno()
	  o Add function to get/set peer pid
	  o Add IFF_LOWER_UP
	  o Add/export some missing accessor functions
	  o print /0 prefix in nl_addr2str()
	  o Fix invalid free in nl_addr_parse for AF_UNSPEC addresses
	  o Use __str2flags instead of __str2type in rtnl_link_str2flags()
	  o Make sure object and filter types match in nl_object_match()
	  o Add support for credential passing over netlink sockets (API BREAK)
	  o Add support for custom dump callbacks
	  o Add NL_DUMP_ENV format

	Michael Biebl <biebl@teco.edu>
	"Alex V. Myltsev" <avm@altlinux.ru>
	  o Makefile fixes


Summary of Changes from 1.0-pre4 to 1.0-pre5
================================================
	Thomas Graf <tgraf@suug.ch>

src/genl-ctrl-list.c
/*
 * src/genl-ctrl-list.c	List Generic Netlink Controller
 *
 *	This library is free software; you can redistribute it and/or
 *	modify it under the terms of the GNU Lesser General Public
 *	License as published by the Free Software Foundation version 2.1
 *	of the License.
 *
 * Copyright (c) 2003-2009 Thomas Graf <tgraf@suug.ch>
 */

#include <netlink/cli/utils.h>

static struct nl_cache *alloc_genl_family_cache(struct nl_sock *sk)
{
	return nl_cli_alloc_cache(sk, "generic netlink family",
			   genl_ctrl_alloc_cache);
}

static void print_usage(void)
{
	printf(
	"Usage: genl-ctrl-list [--details]\n"
	"\n"
	"Options\n"
	" -d, --details         Include detailed information in the list\n"
	" -h, --help            Show this help\n"
	" -v, --version         Show versioning information\n"
	);
	exit(0);
}

int main(int argc, char *argv[])
{
	struct nl_sock *sock;
	struct nl_cache *family_cache;
	struct nl_dump_params params = {
		.dp_type = NL_DUMP_LINE,
		.dp_fd = stdout,
	};
 
	sock = nl_cli_alloc_socket();
	nl_cli_connect(sock, NETLINK_GENERIC);
	family_cache = alloc_genl_family_cache(sock);
 
	for (;;) {
		int c, optidx = 0;
		static struct option long_opts[] = {
			{ "details", 0, 0, 'd' },
			{ "format", 1, 0, 'f' },
			{ "help", 0, 0, 'h' },
			{ "version", 0, 0, 'v' },
			{ 0, 0, 0, 0 }
		};
	
		c = getopt_long(argc, argv, "df:hv", long_opts, &optidx);
		if (c == -1)
			break;

		switch (c) {
		case 'f': params.dp_type = nl_cli_parse_dumptype(optarg); break;
		case 'd': params.dp_type = NL_DUMP_DETAILS; break;
		case 'h': print_usage(); break;
		case 'v': nl_cli_print_version(); break;
		}
 	}

	nl_cache_dump(family_cache, &params);

	return 0;
}
